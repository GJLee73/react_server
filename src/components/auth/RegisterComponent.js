import React from 'react';
import {Link} from 'react-router-dom';

class RegisterComponent extends React.Component {
	constructor(props){
		super(props);
		this.state = {
			userId:"",
            password:"",
            pwdVerifyPhrase:'비밀번호는 8자이상 최소 한개 이상의 특수문자가 있어야합니다.',
            pwdVerified:false,
            pwdCheckPhrase:'',
			passwordCheck:"",
			pwdCheck:false,
			name:'',
			nickname:'',
			birth:'',
			gender:'',
			email:'',
			phone:''
		};
		this.handleChange = this.handleChange.bind(this);
		this.handleRegister = this.handleRegister.bind(this);
        this.checkPwd=this.checkPwd.bind(this);
        this.verify=this.verify.bind(this)
	}

	handleRegister(){
		var msg = {
			userId:this.state.userId,
            password:this.state.password,
			name:this.state.name,
			nickname:this.state.nickname,
			birth:this.state.birth,
			gender:this.state.gender,
			email:this.state.email,
			phone:this.state.phone
		};
		if(!this.props.idCheck){ //alert창 이쁘게 띄울 수 있는지..
		
			return alert('id를 확인하세요')
		}else if(!this.state.pwdVerified||!this.state.pwdCheck){
			return alert('비밀번호를 확인하세요')
		}else if(!this.props.emailCheck){
			return alert('이메일을 확인하세요')
		}
		for(var p in msg){ 
			if(msg.p===''){
				return alert(p,"가 비어있습니다.")
			}
		}
		this.props.onRegister(msg)
			
		
	}

	handleChange(e){ 
		let nextState = {};
		nextState[e.target.name] = e.target.value;
		this.setState(nextState);
	}

	/**
	 * 패스워드가 동일한지 체크함
	 */
    checkPwd(){
		if(this.state.password !==this.state.passwordCheck || !this.state.password || !this.state.passwordCheck)
		{
			return this.setState({pwdCheckPhrase:'비밀번호가 다릅니다.'})
		}
		this.setState({pwdCheckPhrase:'비밀번호가 같습니다.',pwdCheck:true});
    }
    /** 비밀번호가 올바른지 체크한 후 현재 비밀번호 상태를 업데이트한다.
     * @param {string} e -password
    */
    verify(e){
        var stringRegx = /[!@#$%^&*]/gi; 
        if(e.target.value.length>2){
            if(!stringRegx.test(e.target.value)){
                return this.setState({pwdVerifyPhrase:'최소 한개 이상의 특수문자가 포함되어야 합니다.(!@#$%^&*)'})
            }else if(e.target.value.length<8){
                return this.setState({pwdVerifyPhrase:'비밀번호가 너무 짧습니다.'})
            }
            this.setState({pwdVerifyPhrase:'안전한 비밀번호입니다.',pwdVerified:true})
        }  
       
    }
	render(){
		
            

		return (
			<div className="container auth">
				<Link className="logo" to="/">5437</Link>
				<div className="card">
					<div className="header blue white-text center">
						<div className="card-content">REGISTER</div>
					</div>
					<div>
                 <div className="card-content">
                <div className="row">
                <div className="input-field col s12 username">
                    <label>Id</label>
					<i>{this.props.idStatus}</i>
                    <input
                    name="userId"
                    type="text"
                    className="validate"
                    onChange={this.handleChange}
                    value={this.state.userId}
                    onBlur={(e)=>{this.props.checkId(e.target.value)}}/>
                </div>
                <div className="input-field col s12">
                    <label>Password </label>
                    <i>{this.state.pwdVerifyPhrase}</i>
                    <input
                    name="password"
                    type="password"
                    className="validate"
                    onChange={this.handleChange}
                    value={this.state.password}
                    onKeyUp={this.verify}/>
                    </div>

					<div className="input-field col s12" >
                        <label>Passsword check</label>
                        <i>{this.state.pwdCheckPhrase}</i>
                        <input
                        name="passwordCheck"
                        type="password"
                        className="validate"
                        onChange={this.handleChange}
                        value={this.state.passwordCheck}
                        onKeyUp={this.checkPwd}/>
               		</div> 
					<div className="input-field col s12 ">
						<label>이름</label>
						<input
						name="name"
						type="text"
						className="validate"
						onChange={this.handleChange}
						value={this.state.name}/>
               		</div>
					<div className="input-field col s12 ">
						<label>닉네임</label>
						<input
						name="nickname"
						type="text"
						className="validate"
						onChange={this.handleChange}
						value={this.state.nickname}/>
               		</div>
					<div className="input-field col s12 ">
						<label>생년월일</label>
						<input
						name="birth"
						type="text"
						className="validate"
						onChange={this.handleChange}
						value={this.state.birth}/>
               		</div>
					<div className="input-field col s12 ">
					<label>성별..</label>
						<select name='gender' value={this.state.gender} onChange={this.handleChange}>
						<option value="" disabled selected>성별</option>
						<option value="M">남자</option>
						<option value="F">여자</option>
						</select>
               		</div>
					   <div class="input-field col s12">
    <select>
      <option value="" disabled selected>Choose your option</option>
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
    </select>
    <label>Materialize Select</label>
  </div>
  <div class="input-field col s12">
    <select>
      <option value="" disabled selected>Choose your option</option>
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
    </select>
    <label>Materialize Select</label>
  </div>
					<div className="input-field col s6 ">
						<label>이메일</label>
						<i>{this.props.emailStatus}</i>
						<input
						name="email"
						type="text"
						className="validate"
						onChange={this.handleChange}
						value={this.state.email}
						/>
               		</div>
					<div className="input-field col s6 ">
					<label></label>
						<button onClick={()=>{this.props.checkEmail(this.state.email)}}>이메일체크</button>
					
               		</div>
					<div className="input-field col s6 ">
						<label>휴대폰</label>
						<input
						name="phone"
						type="text"
						className="validate"
						onChange={this.handleChange}
						value={this.state.phone}/>
               		</div>
					   <div className="input-field col s6 ">
					   <label></label>
						<button onClick={(e)=>{this.props.checkEmail(e.target.value)}}>폰번호체크</button>
					
               		</div>
                    <a className="waves-effect waves-light btn" onClick={this.handleRegister}>CREATE</a>
                	
                </div>
            </div>
	
				</div>
			</div>
            </div>
		);
	}
}

export default RegisterComponent;