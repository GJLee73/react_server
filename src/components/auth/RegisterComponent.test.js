import React from 'react';
import {shallow,mount} from 'enzyme'
import RegisterComponent from './RegisterComponent'
import {MemoryRouter} from 'react-router-dom'

describe('RegisterComponent Test',()=>{
    var handleChange = jest.spyOn(RegisterComponent.prototype,'handleChange')
    var handleRegister = jest.spyOn(RegisterComponent.prototype,'handleRegister')
    var checkPwd = jest.spyOn(RegisterComponent.prototype,'checkPwd')
    var verify = jest.spyOn(RegisterComponent.prototype,'verify')
    var autoHypenPhone = jest.spyOn(RegisterComponent.prototype,'autoHypenPhone')
    var handleVerifyEmail = jest.spyOn(RegisterComponent.prototype,'handleVerifyEmail')
    var onRegister = jest.fn()
    var checkId = jest.fn()
    var checkEmail = jest.fn()
    var checkPhone = jest.fn()
    global.alert = jest.fn()
    
    var wrapper = shallow(<RegisterComponent idCheck={false} idStatus='hdifhih' onRegister={onRegister} checkId={checkId}
                                                checkEmail={checkEmail} checkPhone={checkPhone}/>)

    var idEvent= {target:{value:'g1',name:'userId'}}
    var pwdEvent={target:{value:'54375437!', name:'password'}}
    var passwordCheck={target:{value:'54375437!', name:'passwordCheck'}}
    var name={target:{value:'김지원',name:'name'}}
    var birth={target:{value:'950613',name:'birth'}}
    var gender={target:{value:'F',name:'gender'}}
    var email={target:{value:'jwc2094@naver.com',name:'email'}}
    var phone={target:{value:'000-0000-0000',name:'phone'}}

    var inputs=wrapper.find('input')
    describe('function test',async ()=>{
        afterEach(()=>{

            handleChange.mockRestore();
        })
        describe('handleChange should work',()=>{
            it('should work well',async()=>{
                expect(handleChange.mock.calls.length).toBe(0)
                inputs.at(0).simulate('change',idEvent)
                inputs.at(1).simulate('change',pwdEvent)
                inputs.at(2).simulate('change',passwordCheck)//pwdcheck
                inputs.at(3).simulate('change',name)
                inputs.at(4).simulate('change',birth)
                // inputs.at(5).simulate('change',gender) 이거는 select
                inputs.at(5).simulate('change',email)
                wrapper.find('select').simulate('change',gender)
                // inputs.at(7).simulate('change',phone) autoHypenPhone 사용.
                await expect(handleChange.mock.calls.length).toBe(7)
                expect(wrapper.state()).toEqual({"birth": "950613", "email": "jwc2094@naver.com", "gender": "F", "name": "김지원","password": "54375437!", "passwordCheck": "54375437!", "phone": "", "pwdCheck": false, "pwdCheckPhrase": "", "pwdVerified": false, "pwdVerifyPhrase": "비밀번호는 8자이상 최소 한개 이상의 특수문자가 있어야합니다.", "userId": "g1"})
        })
        })
        
        describe('checkPwd test',()=>{
            it('비밀번호가 다릅니다.',()=>{
                inputs.at(2).simulate('change',{target:{value:'5437',name:'passwordCheck'}})
                inputs.at(2).simulate('keyup')
                expect(wrapper.state().pwdCheck).toBe(false)
                expect(wrapper.state().pwdCheckPhrase).toBe('비밀번호가 다릅니다.')
            })
            it('비밀번호가 같습니다.',()=>{
                inputs.at(2).simulate('change',{target:{value:'54375437!',name:'passwordCheck'}})
                inputs.at(2).simulate('keyup')
                expect(wrapper.state().pwdCheck).toBe(true)
                expect(wrapper.state().pwdCheckPhrase).toBe('비밀번호가 같습니다.')
            })
           

        })
        describe('pwd verify test',()=>{
            it('length<3',async()=>{
                await inputs.at(1).simulate('change',{target:{value:'1',name:'password'}})
                await inputs.at(1).simulate('keyup',{target:{value:'1',name:'password'}})
                await expect(wrapper.state().pwdVerifyPhrase).toBe("비밀번호는 8자이상 최소 한개 이상의 특수문자가 있어야합니다.")
               })
            it('2<length&&특수문자가 없을때',async()=>{
                await inputs.at(1).simulate('change',{target:{value:'5437',name:'password'}})
                await inputs.at(1).simulate('keyup',{target:{value:'5437',name:'password'}})
                expect(wrapper.state().pwdVerifyPhrase).toBe('최소 한개 이상의 특수문자가 포함되어야 합니다.(!@#$%^&*)')
            })
            it('2<length<8 &&특수문자가 있을때',async()=>{
                await inputs.at(1).simulate('change',{target:{value:'5437!',name:'password'}})
                await inputs.at(1).simulate('keyup',{target:{value:'5437!',name:'password'}})
                expect(wrapper.state().pwdVerifyPhrase).toBe('비밀번호가 너무 짧습니다.')
            })
            it('lenth>8 && 특수문자 포함',async()=>{
                await inputs.at(1).simulate('change',{target:{value:'54375437!',name:'password'}})
                await inputs.at(1).simulate('keyup',{target:{value:'54375437!',name:'password'}})
                expect(wrapper.state().pwdVerifyPhrase).toBe('안전한 비밀번호입니다.')
                expect(wrapper.state().pwdVerified).toBe(true)
            })
        })
        describe('autoHypenPhone test',()=>{
            it('lenth<4',()=>{
                inputs.at(6).simulate('change',{target:{value:'000',name:'phone'}})
                expect(wrapper.state().phone).toBe('000')
            })
            it('4<=lenth<7',()=>{
                inputs.at(6).simulate('change',{target:{value:'00000',name:'phone'}})
                expect(wrapper.state().phone).toBe('000-00')
            })
            it('7<=lenth<11',()=>{
                inputs.at(6).simulate('change',{target:{value:'00000000',name:'phone'}})
                expect(wrapper.state().phone).toBe('000-000-00')
            })
            it('length=11',()=>{
                inputs.at(6).simulate('change',{target:{value:'00000000000',name:'phone'}})
                expect(wrapper.state().phone).toBe('000-0000-0000')
            })
        })
        describe('handleVerifyEmail test',()=>{
            it('이메일 양식이 틀릴때',()=>{
                wrapper.setState({email:'enen'})
                wrapper.find('button').at(0).simulate('click')
                expect(global.alert.mock.calls[0][0]).toBe('올바른 이메일을 입력하세요!')
            })
            it('이메일 양식이 맞을때',()=>{
                wrapper.setState({email:'jwc2094@naver.com'})
                wrapper.find('button').at(0).simulate('click')
                expect(global.alert).toHaveBeenCalledTimes(1)
                expect(checkEmail.mock.calls.length).toBe(1)
            })
        })
        describe('handleRegister test',()=>{

                it('id check fail',()=>{          
                    wrapper.find('a').simulate('click')
                    expect(global.alert).toHaveBeenCalledTimes(2)
                    expect(onRegister).toHaveBeenCalledTimes(0)
                })
                it('not pwd verified',async()=>{
                    wrapper.setProps({idCheck:true}) 
                    wrapper.setState({pwdVerified:false})
                    wrapper.find('a').simulate('click')
                    expect(global.alert.mock.calls[2][0]).toBe('비밀번호를 확인하세요')
                    expect(onRegister).toHaveBeenCalledTimes(0)
                })
                it('not pwdChecked',()=>{
                    wrapper.setProps({idCheck:true})
                    wrapper.setState({pwdVerified:true, pwdCheck:false})
                    wrapper.find('a').simulate('click')
                    expect(global.alert.mock.calls[3][0]).toBe('비밀번호를 확인하세요')
                    expect(onRegister).toHaveBeenCalledTimes(0)
                })
                it('not emailCheck',()=>{
                    wrapper.setProps({idCheck:true,emailCheck:false})
                    wrapper.setState({pwdVerified:true, pwdCheck:true})
                    wrapper.find('a').simulate('click')
                    expect(global.alert.mock.calls[4][0]).toBe('이메일을 확인하세요')
                    expect(onRegister).toHaveBeenCalledTimes(0)
                })
                it('not phoneCheck',()=>{
                    wrapper.setProps({idCheck:true,emailCheck:true,phoneCheck:false})
                    wrapper.setState({pwdVerified:true, pwdCheck:true})
                    wrapper.find('a').simulate('click')
                    expect(global.alert.mock.calls[5][0]).toBe('폰 번호를 확인하세요')
                    expect(onRegister).toHaveBeenCalledTimes(0)
                })
                it('has a blank input',()=>{
                    wrapper.setProps({idCheck:true,emailCheck:true,phoneCheck:true})
                    wrapper.setState({pwdVerified:true, pwdCheck:true, name:''})
                    wrapper.find('a').simulate('click')
                    expect(global.alert.mock.calls[6][0]).toBe('name')
                    expect(onRegister).toHaveBeenCalledTimes(0)
                })
                it('handleRogin Success',()=>{
                    wrapper.setProps({idCheck:true,emailCheck:true,phoneCheck:true})
                    wrapper.setState({pwdVerified:true, pwdCheck:true, name:'g1'})
                    wrapper.find('a').simulate('click')
                    expect(onRegister).toHaveBeenCalledTimes(1)
                })
        })
        
        describe('props function test',()=>{
            it('checkId should work',()=>{
                inputs.at(0).simulate('blur',{target:{value:'g1'}})
                expect(checkId).toHaveBeenCalledTimes(1)
            })
            it('checkPhone should work',()=>{
                wrapper.find('button').at(1).simulate('click')
                expect(checkPhone).toHaveBeenCalledTimes(1)
            })
           
        })
    })
   
    it('snapShot test',()=>{
        expect(wrapper).toMatchSnapshot()
    })

    
})