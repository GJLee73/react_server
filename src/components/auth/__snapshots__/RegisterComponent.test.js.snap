// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterComponent Test snapShot test 1`] = `
<div
  className="container auth"
>
  <Link
    className="logo"
    replace={false}
    to="/"
  >
    5437
  </Link>
  <div
    className="card"
  >
    <div
      className="header blue white-text center"
    >
      <div
        className="card-content"
      >
        REGISTER
      </div>
    </div>
    <div>
      <div
        className="card-content"
      >
        <div
          className="row"
        >
          <div
            className="input-field col s12 username"
          >
            <label>
              Id
            </label>
            <i>
              hdifhih
            </i>
            <input
              className="validate"
              name="userId"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="g1"
            />
          </div>
          <div
            className="input-field col s12"
          >
            <label>
              Password 
            </label>
            <i>
              안전한 비밀번호입니다.
            </i>
            <input
              className="validate"
              name="password"
              onChange={[Function]}
              onKeyUp={[Function]}
              type="password"
              value="54375437!"
            />
          </div>
          <div
            className="input-field col s12"
          >
            <label>
              Passsword check
            </label>
            <i>
              비밀번호가 같습니다.
            </i>
            <input
              className="validate"
              name="passwordCheck"
              onChange={[Function]}
              onKeyUp={[Function]}
              type="password"
              value="54375437!"
            />
          </div>
          <div
            className="input-field col s12 "
          >
            <label>
              이름
            </label>
            <input
              className="validate"
              name="name"
              onChange={[Function]}
              type="text"
              value="g1"
            />
          </div>
          <div
            className="input-field col s12 "
          >
            <label>
              생년월일
            </label>
            <input
              className="validate"
              name="birth"
              onChange={[Function]}
              type="text"
              value="950613"
            />
          </div>
          <div
            className="input-field col s12 "
          >
            <select
              name="gender"
              onChange={[Function]}
              value="F"
            >
              <option
                disabled={true}
                value=""
              >
                성별
              </option>
              <option
                value="M"
              >
                남자
              </option>
              <option
                value="F"
              >
                여자
              </option>
            </select>
          </div>
          <div
            className="input-field col s12 "
          >
            <label>
              이메일
            </label>
            <i />
            <input
              className="validate"
              name="email"
              onChange={[Function]}
              type="text"
              value="jwc2094@naver.com"
            />
          </div>
          <div
            className="input-field col s12 "
          >
            <label />
            <button
              onClick={[Function]}
            >
              이메일체크
            </button>
          </div>
          <div
            className="input-field col s12 "
          >
            <label>
              휴대폰
            </label>
            <input
              className="validate"
              maxLength="13"
              name="phone"
              onChange={[Function]}
              type="text"
              value="000-0000-0000"
            />
          </div>
          <div
            className="input-field col s12 "
          >
            <label />
            <button
              onClick={[Function]}
            >
              폰번호체크
            </button>
          </div>
          <a
            className="waves-effect waves-light btn"
            onClick={[Function]}
          >
            CREATE
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
